{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f06f558b-d98a-4425-b990-46713f6ed53e",
   "metadata": {},
   "source": [
    "# Tutorial 2: Indexing assemblies (SARS-CoV-2 dataset)\n",
    "\n",
    "# 1. Background\n",
    "\n",
    "This tutorial shows you how to construct an index of assembled genomes for querying and visualization. The data consists of a set of 50 SARS-CoV-2 genomes downloaded from the NCBI Virus portal (https://www.ncbi.nlm.nih.gov/labs/virus/) on May 31, 2021.\n",
    "\n",
    "# 2. Getting data\n",
    "\n",
    "Let's first download the data for this tutorial. To do this please run the below commands:\n",
    "\n",
    "*Note: In a Jupyter Python notebook, prepending a command with `!` runs the command in a shell instead of the Python interpreter (e.g., `!unzip` runs the command `unzip`).*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73c22f2e-0f52-469b-8119-e529f8b2ddff",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O sars-cov-2-genbank-50.zip https://ndownloader.figshare.com/files/29095476?private_link=e54aded21bb036071aac\n",
    "!unzip -n sars-cov-2-genbank-50.zip | head -n 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da28c89-6435-42a4-968d-a082cd8f3aa6",
   "metadata": {},
   "source": [
    "## 2.1. Examine data\n",
    "\n",
    "Let's look at what files are available in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca24edc-3bde-4793-a203-1dacc110ccdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls -F sars-cov-2-genbank-50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b271eb9-246e-44c2-9060-d5af002beb71",
   "metadata": {},
   "source": [
    "There is a file of metadata `sars-cov-2.metadata.csv` (from NCBI) which looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b40484c8-4d85-4d50-9cfd-84265f0f82b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.read_csv('sars-cov-2-genbank-50/sars-cov-2.metadata.csv').head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9725c3b-a447-447f-88c5-ffe7cba11749",
   "metadata": {},
   "source": [
    "There's a reference genome in the `reference/` directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab4e29b6-73d8-49ea-a54b-21a214a1816e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls sars-cov-2-genbank-50/reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d0ed1f3-e284-4c3c-8cc8-8312712b5fd6",
   "metadata": {},
   "source": [
    "And, there are the input SARS-CoV-2 genomes in the `fasta/` directory, one genome per file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47451196-db24-460a-b09d-4fb4fcc03daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls sars-cov-2-genbank-50/fasta | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a315eb4d-ad8f-465b-b4b5-9c9035fbf663",
   "metadata": {},
   "source": [
    "Now that we've seen the input data, let's go through generating an index.\n",
    "\n",
    "# 3. Create an index\n",
    "\n",
    "Before we create an index, let's first look at which version of `gdi` we are using."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19df6d1d-2607-433a-b993-a67c9bf8f100",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f274a7e2-7436-4370-8201-d7d8e92bc21e",
   "metadata": {},
   "source": [
    "## 3.1. Initialize an index\n",
    "\n",
    "To create an index, we use the command `gdi init [project]` to construct an project/index directory where we can then load up data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "211ab0f9-9eaa-4a59-8423-92f44eccb71f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi init sars-cov-2-index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305e1109-ffb9-4f8d-9e86-ee2067b76e61",
   "metadata": {},
   "source": [
    "This will create a directory `sars-cov-2-index` which contains the empty index. If we look inside this directory we can see:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0df0bb5-2938-4910-bdfd-0b4277c15a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls -a sars-cov-2-index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "746220d5-6d7e-4d6e-aba1-6fa60460b57d",
   "metadata": {},
   "source": [
    "We see a config file `gdi-config.yaml` which defines some configuration settings:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c811f2-3448-4a7c-a7e8-3d419ba3c4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat sars-cov-2-index/gdi-config.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972bc2ab-1b1a-4443-8494-4cf89581ae2e",
   "metadata": {},
   "source": [
    "Specifically, this file defines the data directory which is used to store VCF/mash/other files. As well as a SQLite database connection path.\n",
    "\n",
    "The `.gdi-data` directory will initially be empty, but will be filled with data as we load data into the index.\n",
    "\n",
    "*Note: you can change from SQLite to another database by removing `sqlite_database=.gdi-data/gdi-db.sqlite` and adding `database_connection=dialect+driver://username:password@host:port/database`, where `dialect+driver...` refers to a [SQLAlchemy-compatible connection string](https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls).*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d4b9b77-eb0e-4dc3-844f-ddddecbbe86e",
   "metadata": {},
   "source": [
    "## 3.2. Index mutations and kmers\n",
    "\n",
    "To index assembled genomes we can use the command `gdi --project-dir sars-cov-2-index analysis [...]`.\n",
    "\n",
    "The project directory must be specified with `--project-dir`. Or, alternatively, you can `cd sars-cov-2-index` and then run `gdi analysis [...]` and leave out the `--project-dir`.\n",
    "\n",
    "Indexing will align the assembled genomes to some reference genome using [minimap2](https://github.com/lh3/minimap2) (for identifying mutations in VCF format) and will also run [sourmash](https://sourmash.readthedocs.io/) for construcing kmer sketches. This is all scheduled using a [Snakemake](https://snakemake.readthedocs.io/) workflow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0cf1ce7-5912-4c97-b89b-ad3125c850aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index analysis \\\n",
    "    --reference-file sars-cov-2-genbank-50/reference/NC_045512.gbk.gz \\\n",
    "    --use-conda \\\n",
    "    --include-kmer \\\n",
    "    --kmer-size 31 --kmer-size 51 --kmer-size 71 \\\n",
    "    --kmer-scaled 100 \\\n",
    "    sars-cov-2-genbank-50/fasta/*.fasta.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25107694-837b-4cd9-980e-305d8a614879",
   "metadata": {},
   "source": [
    "The different parameters are interpreted as follows:\n",
    "    \n",
    "* `--reference-file`: Path to the reference file used for alignment. If you wish to use amino-acids for variant identifiers please specify a Genbank file with gene annotations to use with **snpeff**.\n",
    "* `--use-conda`: Install workflow dependencies using conda\n",
    "* `--include-kmer`: Include kmer (sourmash) sketches.\n",
    "* `--kmer-size`: Specify the kmer-sizes for sourmash.\n",
    "* `--kmer-scaled`: Specify the scaling factor for sourmash.\n",
    "* `sars-cov-2-genbank-50/fasta/*.fasta.gz`: The input genomes to index (assembled genomes).\n",
    "\n",
    "By default, this will use the maximum number of cores available on your machine for processing. This can be adjusted with `gdi --ncores [ncores]`.\n",
    "\n",
    "Note, the slower part of this will likely be installing the tools using conda/mamba.\n",
    "\n",
    "When everything is completed, the index directory `sars-cov-2-index/` should now have data within it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e2aa202-32ad-4070-b706-4a55f5fc685c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls sars-cov-2-index/.gdi-data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef49e812-7ba4-4096-9797-098e7ebb4881",
   "metadata": {},
   "source": [
    "## 3.3. Build phylogenetic tree\n",
    "\n",
    "In addition to indexing mutations or kmer sketches, a phylogenetic tree can be built and saved using all genomes aligned to some specific reference genome. This can be accomplished with the `gdi rebuild tree` command.\n",
    "\n",
    "This command requires us to specify the reference genome name. This is derived from the file defined above, but to view all reference genome names you can use the `gdi list genomes` command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7be9d9a-561f-4c26-8b40-2abca8874a70",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index list genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024ad8f3-7078-48dd-a447-a95e93834a6a",
   "metadata": {},
   "source": [
    "Now that we have a reference genome, let's build a tree. This uses [iqtree](http://www.iqtree.org/) by default.\n",
    "\n",
    "*Note: if you wish you can skip this step and instead you can use the provided tree in `sars-cov-2-genbank-50/tree.newick` when querying. You will have to remember to join a tree to a query with [query.join_tree()](https://github.com/apetkau/genomics-data-index/blob/development/genomics_data_index/api/query/SamplesQuery.py#L177-L188).*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264dc952-f52e-4b23-91f4-6d699cb97acb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index/ rebuild tree --align-type full --extra-params '--fast -m GTR+F+R4' NC_045512"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "928072e1-3471-488a-9e3f-e1ee186fba88",
   "metadata": {},
   "source": [
    "The parameters are interpreted as follows:\n",
    "\n",
    "* `--align-type full`: Use `full` to include all positions on the reference genome (29,903 bp in this case). Use `core` to include only core SNV-containing positions. Note that the tree is built using only SNVs (SNPs), indels or complex variants/mutations are excluded.\n",
    "* `--extra-params`: Extra parameters to pass to `iqtree`. I am specifying the model here to speed up building a tree. Specifying extra parameters is optional.\n",
    "* `NC_045512`: The name of the reference genome to use for building a tree.\n",
    "\n",
    "Once everything is complete you can export the tree in newick format as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b0a1a30-a2f9-40e4-83ea-ad5401ea7228",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index export tree NC_045512"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35131c2f-86c3-4f6c-ac12-dcab44c5e893",
   "metadata": {},
   "source": [
    "# 4. Basic querying/visualization\n",
    "\n",
    "Once you have the index setup you can do some basic querying as follows.\n",
    "\n",
    "I include this here to show a complete tutorial of both indexing and working with data. For an additional querying tutorial please see [Tutorial 1](https://github.com/apetkau/genomics-data-index/blob/development/docs/tutorial/tutorial-1-salmonella.ipynb).\n",
    "\n",
    "## 4.1. Command-line interface\n",
    "\n",
    "### 4.1.1. List samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6b26b76-0d21-452c-b88f-7fde8c790b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index list samples | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed5adaac-54f2-4afa-bbd2-0c61aa0b5223",
   "metadata": {},
   "source": [
    "### 4.1.2. List reference genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "428f6f4f-c566-4f61-b5e0-7cc24ee7b80e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gdi --project-dir sars-cov-2-index list genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "854010cb-cfe5-495c-87a1-1deb8eefa1e3",
   "metadata": {},
   "source": [
    "## 4.2. Python API\n",
    "\n",
    "First lets connect to the index in the directory `sars-cov-2-index`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054ad822-d8fc-4a3c-b8de-01a5ef04ce42",
   "metadata": {},
   "outputs": [],
   "source": [
    "import genomics_data_index.api as gdi\n",
    "\n",
    "db = gdi.GenomicsDataIndex.connect('sars-cov-2-index')\n",
    "db"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c1ce4e6-b98d-4d56-b57c-c2af5ff30845",
   "metadata": {},
   "source": [
    "Now let's load up the dataframe of provided metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f68304-d278-4dc7-a7fc-64de44c07089",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv('sars-cov-2-genbank-50/sars-cov-2.metadata.csv')\n",
    "metadata_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22b672a1-d843-473b-a264-9cf476065527",
   "metadata": {},
   "source": [
    "### 4.2.1. Create query which includes tree\n",
    "\n",
    "To create a query which includes a phylogenetic tree you will have to specify the reference genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0913411e-dd4a-4b99-b38c-deb701825289",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = db.samples_query(universe='mutations', reference_name='NC_045512')\n",
    "q"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ae549d0-0947-4a1c-a0eb-7aa792beac91",
   "metadata": {},
   "source": [
    "### 4.2.2. Create query and join tree from file\n",
    "\n",
    "Alternatively, you can create a query and join to a tree stored in a separate file or some other external resource.\n",
    "\n",
    "First you have to load the tree using the **ETEToolkit**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b088c905-7690-4842-a1a6-e37fad1751f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "tree = Tree('sars-cov-2-genbank-50/tree.newick')\n",
    "tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "249df10f-0182-41eb-b58e-954f83d3b7e6",
   "metadata": {},
   "source": [
    "Now we join this tree to a query. You have to specify the alignment length used to construct the tree (in this case the same as the length of the reference genome). This is so that I can do conversions between distance units (**substitutions/site** to **substitutions**).\n",
    "\n",
    "*Note: This may not quite be working 100% yet. This is still in-development software.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a2851d-32ee-4b49-8ce3-aef21ed30bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = db.samples_query().join_tree(tree, alignment_length=29903)\n",
    "q"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a749481d-4df8-448c-97e4-275ef932a453",
   "metadata": {},
   "source": [
    "### 4.2.3. Create query with tree and dataframe of metadata\n",
    "\n",
    "To create a query with both the metadata and a tree attached we can do the following."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3682ef-654c-4335-9724-a1f790f59fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = db.samples_query(\n",
    "    universe='mutations', reference_name='NC_045512').join(metadata_df, sample_names_column='Accession')\n",
    "q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9cc0361-a5a7-4f32-af1b-78d5e4341ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set outgroup to make visualizations look a bit better\n",
    "# I would like to find a better way of specifying the outgroup instead of\n",
    "# directly calling the ete3 API here\n",
    "q.tree.set_outgroup('NC_045512')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b8147f-b0a2-488d-b9f2-652480c7ae23",
   "metadata": {},
   "source": [
    "### 4.2.4. Show summary of mutations\n",
    "\n",
    "This shows a summary of all mutations accross all genomes in the query. The **ID_HGVS...** columns show identifiers using [HGVS](https://varnomen.hgvs.org/) as derived from `snpeff`. You can use each of these identifiers (or the **Mutation** column) to search for particular mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e5f756-c786-4bfe-9cc7-c2bfd73b096b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = q.features_summary().sort_values(\n",
    "    'Count', ascending=False).reset_index().set_index('ID_HGVS_GN.p', drop=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f9791bc-a41c-438e-84cd-f32688e77594",
   "metadata": {},
   "source": [
    "### 4.2.5. Find genomes with a particular mutation\n",
    "\n",
    "This shows a summary of all genomes with a particular mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d2b1160-f0f6-47bc-8f7c-a80694863735",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.hasa('hgvs_gn:NC_045512.2:S:p.D614G').summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157bab61-d90b-46a4-aeb1-57869679ff89",
   "metadata": {},
   "source": [
    "To view all individual genomes with this mutation, you can use `.toframe()` (I also use `.head(3)` to not clutter up the notebook with data from all 38 genomes with a `D614G` mutation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b86db40e-13f5-4af5-aa83-51a10894007e",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.hasa('hgvs_gn:NC_045512.2:S:p.D614G').toframe().head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fefc460d-cd2e-4c41-a74a-324d29032a69",
   "metadata": {},
   "source": [
    "### 4.2.6. Visualize queries on tree\n",
    "\n",
    "Here, we can visualize queries on a phylogenetic tree. In this case, we are looking for genomes with particular mutations `S:p.D614G` and `ORF8:p.L84S`.\n",
    "\n",
    "The boxes shown next to the tree include three categories: **[mutation]** (colored black) if the mutation is present, **U** (colored gray) if it is unknown if the mutaiton exists, and **[nothing]** (colored white) if the mutation does not exist. Those genomes for which it is unknown if the mutation exists is defined as those genomes where the mutation overlaps a region with gaps `-` or ambiguous bases `N` in the genome in question. \n",
    "\n",
    "*Note: those genomes where the mutation does not exist is **not** equivalent to saying they have the reference genome amino acid since it's possible there is a different mutation in this location*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e44ea3-6f6e-49ca-958e-5540274b1e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.tree_styler(show_legend_type_labels=False, annotate_show_box_label=True)\\\n",
    "    .annotate(q.hasa('hgvs_gn:NC_045512.2:S:p.D614G'), box_label='S:D614G')\\\n",
    "    .annotate(q.hasa('hgvs_gn:NC_045512.2:ORF8:p.L84S'), box_label='ORF8:L84S')\\\n",
    "    .render(w=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d2fe20-04e7-4f5c-b364-15460edc40b2",
   "metadata": {},
   "source": [
    "### 4.2.7. Highlight lineages\n",
    "\n",
    "To highlight lineages first we show which lineages are available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1f5933-4a4d-4c46-810e-3c374e1169aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.toframe()['Pangolin'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "883d0ca0-e552-4785-9dfc-f52fb2dadffc",
   "metadata": {},
   "source": [
    "We can query for samples that are of a specific lineage with `isa()`.\n",
    "\n",
    "You can use `regex=True` to use regular expressions in the lineage query (to capture that `A.1` is also an `A` and `B.1` is also a `B`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5492906-a289-4bc5-b3d6-6421f219c348",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.isa('^A', regex=True, kind='dataframe', isa_column='Pangolin')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81755894-1e47-4527-bae8-fd464aea8db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.isa('^B', regex=True, kind='dataframe', isa_column='Pangolin')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10a35a2b-0ac3-45b4-8460-c2811ce6532a",
   "metadata": {},
   "source": [
    "Now lets combine with previous visualization to show these lineages in relation to mutations on a tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6370ddb0-318f-41f2-a510-82e2524434f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.tree_styler(show_legend_type_labels=False, annotate_show_box_label=True,\n",
    "             legend_title='Lineage')\\\n",
    "    .annotate(q.hasa('hgvs_gn:NC_045512.2:S:p.D614G'), box_label='S:D614G')\\\n",
    "    .annotate(q.hasa('hgvs_gn:NC_045512.2:ORF8:p.L84S'), box_label='ORF8:L84S')\\\n",
    "    .highlight(q.isa('^A', regex=True, kind='dataframe', isa_column='Pangolin'), legend_label='A')\\\n",
    "    .highlight(q.isa('^B', regex=True, kind='dataframe', isa_column='Pangolin'), legend_label='B')\\\n",
    "    .render(w=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cde349bf-f569-4bd1-9420-3b5744e36076",
   "metadata": {},
   "source": [
    "### 4.2.8. Show summary of mutations for a lineage\n",
    "\n",
    "Let's show a summary of mutations for lineage `AD.2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e924f140-83cd-4337-b580-4662330c5403",
   "metadata": {},
   "outputs": [],
   "source": [
    "q.isa('AD.2', kind='dataframe', isa_column='Pangolin')\\\n",
    " .features_summary()\\\n",
    " .sort_values('Count', ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1be21794-be9b-4846-b1c8-7578b925273f",
   "metadata": {},
   "source": [
    "### 4.2.9. Unique features\n",
    "\n",
    "We can also produce a table of unique features found only in lineage `AD.2` (out of everything in our index)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b08a318c-2c8a-4679-95c7-dd652004ab78",
   "metadata": {},
   "outputs": [],
   "source": [
    "u_df = q.isa('AD.2', kind='dataframe', isa_column='Pangolin').features_summary(selection='unique')\n",
    "u_df.sort_values('Count', ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4592e72a-f283-4e9b-9dd6-a44bf5868cdb",
   "metadata": {},
   "source": [
    "### 4.2.10. Compare feature proportions among lineages\n",
    "\n",
    "Sometimes just showing the unique features for a lineage will miss some important information when some (but very few) genomes from other lineages have a feature. For example, if 100% of genomes in lineage `AD.2` have a mutation, but 1% of genomes in other lineages have this same mutation, it will not show up from the unique features analysis above, but still may be useful to know.\n",
    "\n",
    "In these situations, you can use the `q.features_comparison()` method to generate a table of each lineage (or other division of samples) along with the percent/proportion of every feature found in this lineage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38320ebe-c427-4ed5-b84b-45d647294849",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_df = q.features_comparison(sample_categories='Pangolin',\n",
    "                      categories_kind='dataframe',\n",
    "                      kind='mutations',\n",
    "                      unit='percent'\n",
    "                      )\n",
    "comparison_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "925830b4-1474-4f65-ae37-27f453685fb6",
   "metadata": {},
   "source": [
    "#### Parameters\n",
    "\n",
    "The parameters for this command are:\n",
    "\n",
    "* **sample_categories='Pangolin'**: The column name (`Pangolin`) in the dataframe to define the sets of genomic samples to compare.\n",
    "* **categories_kind='dataframe'**: The kind of categories to compare (`dataframe` means that we are defining sets of samples based on a column name in the attached metadata dataframe. Alternative methods is to pass sample queries to define sets of samples).\n",
    "* **kind='mutations'**: The kind of features to compare (can be `mutations` or `mlst`).\n",
    "* **unit='percent'**: The kind of unit to use (can be `percent`, `proportion`, or `count`).\n",
    "* There are others and I would recommend referring to the Python API documentation for more details (e.g., run `?q.features_comparison` in Jupyter).\n",
    "\n",
    "#### Output\n",
    "\n",
    "The output of this command consists of a dataframe where rows are the features and columns (like `B.1_percent`, `AD.2_percent`, ..., `B.1_total`, `AD.2_total`) list the percent of each mutation for the given lineages (`B.1`, `AD.2`) as well as total samples in each lineage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2b1248-8e80-4a95-8e8b-6bbabee214a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_df[['ID_HGVS_GN.p', 'B.1_percent', 'AD.2_percent', 'B.1_total', 'AD.2_total']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09df04f1-71ae-4005-a879-0c5d352926a0",
   "metadata": {},
   "source": [
    "For example, here you can see that the feature `ORF3a:p.T223I` is found in **100%** (out of 10 samples) of lineage **AD.2** and **0%** (out of 28 samples) in lineage **B.1**.\n",
    "\n",
    "You can even use this information to plot out the major differences in mutations between lineages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa99b392-4f4e-4b37-98f5-e00830b5e996",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "comparison_diff_df = comparison_df.copy()\n",
    "\n",
    "# Add in column which defines the differences between percent of features in\n",
    "# the selected lineages\n",
    "comparison_diff_df['B.1_minus_AD.2_percent'] = comparison_diff_df['B.1_percent'] \\\n",
    "                                               - comparison_diff_df['AD.2_percent']\n",
    "\n",
    "# Use this difference column (B.1_minus_AD.2_percent) to sort features to compare\n",
    "comparison_diff_df = comparison_diff_df.sort_values('B.1_minus_AD.2_percent',\n",
    "                                                    ascending=False)\n",
    "\n",
    "# Limit data to only those features where the difference \n",
    "# between lineages is greater than 10%\n",
    "comparison_diff_df = comparison_diff_df[\n",
    "    comparison_diff_df['B.1_minus_AD.2_percent'] > 10][\n",
    "    ['ID_HGVS_GN.p', 'B.1_percent', 'AD.2_percent']]\n",
    "\n",
    "# Plot percent difference\n",
    "comparison_diff_df.set_index('ID_HGVS_GN.p').plot(kind='barh')\n",
    "plt.xlabel(f'Percent of samples with a particular feature', fontdict={'size': 14})\n",
    "plt.ylabel('Feature', fontdict={'size': 14})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37ce604-dc41-4c07-b832-414586937f55",
   "metadata": {},
   "source": [
    "This shows a plot of the percent of samples with particular features in a lineage. For exmaple, the top entry shows that the feature `N:p.S194L` is found in ~10% of lineage `B.1` and 0% of lineage `AD.2`.\n",
    "\n",
    "I am comparing only two different lineages here (`B.1` and `AD.2`). I have sorted the data so that I display those features with the greatest difference in percent. Nothing appears for lineage `AD.2` since I have limited to showing only features where `B.1` has more than a 10% point difference from `AD.2`. Feel free to adjust the settings in the code above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b30b6ed2-717e-4368-ac1d-5f825c70d43f",
   "metadata": {},
   "source": [
    "### 4.2.11. Build and visualize kmer tree\n",
    "\n",
    "We can also build a kmer-based tree (using the sourmash sketches). We can then visualize this tree and display similar to the mutation-based phylogenetic tree.\n",
    "\n",
    "First we create a query and build a kmer-tree with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73420fc7-711e-4730-a3de-3b6541f0be81",
   "metadata": {},
   "outputs": [],
   "source": [
    "kq = db.samples_query().join(\n",
    "    metadata_df, sample_names_column='Accession').build_tree(\n",
    "    kind='kmer', kmer_size=31)\n",
    "kq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0722c60d-4616-495d-94c4-e99bf015830e",
   "metadata": {},
   "source": [
    "Now let's visualize this tree, showing the same mutations and lineage information as in (**4.2.7**)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9db975ba-5f49-4198-88de-545b4755ec8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "kq.tree_styler(show_legend_type_labels=False, annotate_show_box_label=True,\n",
    "             legend_title='Lineage')\\\n",
    "    .annotate(kq.hasa('hgvs_gn:NC_045512.2:S:p.D614G'), box_label='S:D614G')\\\n",
    "    .annotate(kq.hasa('hgvs_gn:NC_045512.2:ORF8:p.L84S'), box_label='ORF8:L84S')\\\n",
    "    .highlight(kq.isa('^A', regex=True, kind='dataframe', isa_column='Pangolin'), legend_label='A')\\\n",
    "    .highlight(kq.isa('^B', regex=True, kind='dataframe', isa_column='Pangolin'), legend_label='B')\\\n",
    "    .render(w=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9cc8de5-2900-4d37-ab91-ffc00e27c555",
   "metadata": {},
   "source": [
    "*Note: with SARS-CoV-2 being such a small genome where most samples are very closely-related the kmer-based trees will likely not cluster genomes similarly to mutation-based trees (mutation-based trees would be prefered here, but kmer trees may be prefered for more distantly-related genomes).*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4590370-17ef-45b2-b564-02ec5b92128c",
   "metadata": {},
   "source": [
    "# 5. Finished\n",
    "\n",
    "You are now finished with this tutorial. You are awesome 😁❤️. Please feel free to leave issues on GitHub if you encounter any problems."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
